### PR TITLE
`Discussion`: Reduce loading time of discussion page with many posts by using pagination

### PR DESCRIPTION
**Pull request by canberkanar**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German (With translation service, please check). 

### Motivation and Context

Resolves: #4433
### Description
Pagination has been added to course discussion page. All filtering, sorting operations moved to backend server to enhance the performance.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student

1. Log in to Artemis
2. Navigate to Course Discussions Page
3. Verify whether filtering/sorting of posts on page occur as expected.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

AnswerPostService.java 100% | 98%
PostService.java 100% | 78%
ReactionService 100% | 100% 

course-discussion.component.ts: 82.55 %|   95.65%
### Screenshots

image

https://user-images.githubusercontent.com/67344158/153227491-5d35c827-4276-4b03-9ba7-4a62b3c21f94.mp4

### CODE REVIEW

**Review by baranimo7 — APPROVED** [review]
Tested on ts2. Filters worked correctly.

File: src/test/java/de/tum/in/www1/artemis/metis/AnswerPostIntegrationTest.java
Comment (akesfeden) [inline src/test/java/de/tum/in/www1/artemis/metis/AnswerPostIntegrationTest.java:43]:
```suggestion
    private static final int MAX_POSTS_PER_PAGE = 20;
```
This value seems like a constant so it may be better to make it explicit. There are similar usages like that below.

File: src/test/java/de/tum/in/www1/artemis/metis/PostIntegrationTest.java
Comment (akesfeden) [inline src/test/java/de/tum/in/www1/artemis/metis/PostIntegrationTest.java:61]:
```suggestion
    private static final int MAX_POSTS_PER_PAGE = 20;
```
Same as above.

File: src/test/java/de/tum/in/www1/artemis/metis/ReactionIntegrationTest.java
Comment (akesfeden) [inline src/test/java/de/tum/in/www1/artemis/metis/ReactionIntegrationTest.java:55]:
```suggestion
    private static final int MAX_POSTS_PER_PAGE = 20;
```
Same as above.

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java
Comment (akesfeden) [inline src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java:97]: Since it is static, If this value is mutated maybe it would be better to declare it inside the calling test method and pass it as an argument to `createBasicPost` method to prevent the risk of introducing unexpected behavior by having global state, what do you think?

**Review by akesfeden — COMMENTED** [review]
I commented few suggestions.

**Review by b-fein — DISMISSED** [review]
Code review:
+1 for @Strohgelaender’s suggestion to put the filter options into a dedicated class. Maybe the client could also use it and send it with the request to avoid having to convert from `PostContextFilter` to URL parameters on the client and then back again to the class on the server.

Additionally, I left a (quite important, I think) inline comment about the database queries.


You also ticked the ‘I chose a title conforming to the naming conventions for pull requests.’ box in the PR description, but the title does not follow the guidelines.

File: src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java
Comment (canberkanar) [reply src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java]: Thanks for your feedback. The method is called from 3 different locations and my intention was that createBasicPosts is called from anywhere the programmer likes and the posts created have different creation dates. If we update the code like suggested then the caller would have to handle this logic within each function that calls the mentioned method. 

In my opinion, the current implementation is adequate. However, we can change with respect to the severity of the risk.

File: src/test/java/de/tum/in/www1/artemis/metis/AnswerPostIntegrationTest.java
Comment (canberkanar) [reply src/test/java/de/tum/in/www1/artemis/metis/AnswerPostIntegrationTest.java]: Implemented, thanks! 👍🏻

File: src/test/java/de/tum/in/www1/artemis/metis/ReactionIntegrationTest.java
Comment (canberkanar) [reply src/test/java/de/tum/in/www1/artemis/metis/ReactionIntegrationTest.java]: 👍🏻

File: src/main/java/de/tum/in/www1/artemis/repository/metis/PostRepository.java
Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/repository/metis/PostRepository.java]: Thank you for your suggestions. I was mostly able to address the issues except for the sub-queries where I had to use "NOT EXISTS" keyword.  There I have to make sure in a one-to-many relationship that a Post does not contain any answer that is resolved.

File: src/test/java/de/tum/in/www1/artemis/metis/AnswerPostIntegrationTest.java
Comment (canberkanar) [reply src/test/java/de/tum/in/www1/artemis/metis/AnswerPostIntegrationTest.java]: 👍🏻  updated, thanks!

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (willergeorg) [inline src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:326]: Isn't this already done by `preCheckUserAndCourse()`? This seems to be redundant to me.

File: src/test/javascript/spec/component/overview/course-discussion/course-discussion.component.spec.ts
Comment (willergeorg) [inline src/test/javascript/spec/component/overview/course-discussion/course-discussion.component.spec.ts:155]:
```suggestion
        expect(component.formGroup.get('sortBy')?.value).toBe(PostSortCriterion.CREATION_DATE);
        expect(component.formGroup.get('filterToUnresolved')?.value).toBe(false);
        expect(component.formGroup.get('filterToOwn')?.value).toBe(false);
        expect(component.formGroup.get('filterToAnsweredOrReacted')?.value).toBe(false);
```

File: src/test/javascript/spec/component/overview/course-discussion/course-discussion.component.spec.ts
Comment (willergeorg) [inline src/test/javascript/spec/component/overview/course-discussion/course-discussion.component.spec.ts:174]:
```suggestion
        expect(component.currentSortDirection).toBe(SortDirection.DESCENDING);
```

File: src/test/javascript/spec/component/overview/course-discussion/course-discussion.component.spec.ts
Comment (willergeorg) [inline src/test/javascript/spec/component/overview/course-discussion/course-discussion.component.spec.ts:408]: unused variable

**Review by willergeorg — COMMENTED** [review]
Tested on TS1 and worked as expected. I left you some small remarks in the inline comments.

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:326]: The code was implemented multiple times by a previous developer. I noticed that within "preCheckUserAndCourse", checkHasAtLeastRoleInCourseElseThrow is called with user as parameter. In line 326, it is called without a user. We can remove if it can be confirmed that the behavior wont change.

File: src/test/javascript/spec/component/overview/course-discussion/course-discussion.component.spec.ts
Comment (canberkanar) [reply src/test/javascript/spec/component/overview/course-discussion/course-discussion.component.spec.ts]: removed 👍🏻

**Review by akesfeden — DISMISSED** [review]
Tested in the testing session, sorting, pagination, filter and search is working as expected and delegated to server, Codacy warnings are fixed.

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:659]: This pattern is repeated a few times. I think something like
```suggestion
    private static int compareByCreationDate(Post postA, Post postB, SortingOrder sortingOrder) {
        int comparisonResult = 0;
        if (postA.getCreationDate().compareTo(postB.getCreationDate()) > 0) {
            comparisonResult = 1;
        }
        else if (postA.getCreationDate().compareTo(postB.getCreationDate()) < 0) {
            comparisonResult = -1;
        }
        return applySortingOrder(sortingOrder, comparisonResult);
    }

    private static int applySortingOrder(SortingOrder sortingOrder, int comparisonResult) {
        if (SortingOrder.ASCENDING == sortingOrder) {
            return comparisonResult;
        }
        else {
            return -1 * comparisonResult;
        }
    }
```
could reduce the number of branches and the duplication here and in the other helper methods.

Alternatively, the helper methods do not get the the `sortingOrder` parameter at all and instead the `postComparator` saves the relevant `comparisonResult` in a variable and only applies the sort order at the very end.

**Review by b-fein — COMMENTED** [review]
Thanks for implementing the previous suggestions. I added a few more comments where I think the server code can be simplified further.

File: src/main/java/de/tum/in/www1/artemis/repository/metis/PostRepository.java
Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/repository/metis/PostRepository.java:32]: Yes it is actually. For instance, courseWideContext is a boolean parameter which can either be fed in by certain pages to fetch posts according to this column or simply the courseWideContext within the request is sent null to just fetch all posts.

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:610]: I updated the method's name. If you go one method up, you may see that 0 instead of null is returned after all different comparisons if the posts are equal in priority.

**Review by filip-gregurevic — DISMISSED** [review]
Code changes look good to me 👍🏼

**Review by willergeorg — APPROVED** [review]
Tested again on TS2. Still works. Code changes are fine for me as well.

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java]: Yes, but is there some reason why only the full comparison method should adhere to the `Comparable` convention? Wouldn’t it make sense that all helper methods also use this convention of returning `0` if the elements are considered to be equal according to this specific comparison?

Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java]: Yes, I agree with your statement. It makes sense to follow the convention for the sub-methods as well. I updated the code.

**Review by willergeorg — APPROVED** [review]

**Review by filip-gregurevic — DISMISSED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:578]:
```suggestion
        int order = compareByPriority(postA, postB);
```

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:241]:
```suggestion
            postsInCourse = postsInCourse.stream().filter(post -> postFilter(post, postContextFilter.getSearchText())).toList();
```

File: src/test/java/de/tum/in/www1/artemis/metis/ReactionIntegrationTest.java
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/metis/ReactionIntegrationTest.java:240]:
```suggestion
                .comparing((Post post) -> post.getReactions().stream().filter(reaction -> reaction.getEmojiId().equals(VOTE_EMOJI_ID)).count())
```

File: src/test/java/de/tum/in/www1/artemis/metis/ReactionIntegrationTest.java
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/metis/ReactionIntegrationTest.java:278]:
```suggestion
                .comparing((Post post) -> post.getReactions().stream().filter(reaction -> reaction.getEmojiId().equals(VOTE_EMOJI_ID)).count()));
```

**Review by b-fein — COMMENTED** [review]
Thanks for implementing the suggestions, the code looks mostly good now. I’ve got one (probably) final round of suggestions, though.

(Sorry to the quick re-approvers.)

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java]: I started getting a null pointer exception after this change. The behavior of the code above is different. As far as I understood, using .toList(); makes the list uneditable which I must do after this code. You can check the following failing test outputs; https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS3050-JAVATEST-32/test/case/491283038

Comment (canberkanar) [reply src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:240]: Thanks, I revoked this part. You can kindly take a look now. @b-fein

**Review by willergeorg — DISMISSED** [review]
Still good to me.

**Review by dveljacic — COMMENTED** [review]
The pagination/filtering functionality works without any problems. I just have three small comments/questions:

1. Isn't it more common to use 10 results on one page? Currently, it's set to 5, and I believe that in courses with many posts this becomes hard to work with.

2. If there are no posts, no pagination bar should be shown in my opinion. See the following screenshot for reference: 

no_posts

3. This one is more a personal opinion and you can ignore it if you like, but I would insert a break here or padding or something? Currently, for me, it seems too tight and too close to the horizontal bar, and also wording to the pagination bar.

pagination

**Review by akesfeden — APPROVED** [review]
Re-approving after incorporating feedback.

**Review by filip-gregurevic — APPROVED** [review]
Re-approve

**Review by dveljacic — APPROVED** [review]
Retested on TS2 and everything works as expected. Thank you for incorporating my feedback! ✅

**Pull request merged**
